### DISCUSSION TITLE
Add more mypy doc sections on common patterns/pitfalls?

### DISCUSSION THREAD

**Discussion by johnthagen** [collerek/ormar#172] [Ideas]
When selecting inheritance options, I found I had to use **Concrete table inheritance** or else `mypy` would throw type errors when mixins were used as inputs to generic `QuerySet`s:

```py
T = TypeVar("T", bound=MyBaseModel)

def filter(self, query: QuerySet[T], model_class: Type[T]) -> QuerySet[T]:
    # Only works if `T` is an Ormar Model, so can't use Mixins
```

Also, the Python-style filters seem to trip up `mypy`. This is unfortunate, as I feel like they are a bit easier to use and read.

```py
class MyModel(Model):
    length: Optional[str] = Text(nullable=True)


query = MyModel.objects.filter(model_class.length >= length)
# error: Unsupported operand types for <= ("str" and "None")
```

Perhaps these kinds of patterns could be added to the docs with suggested work arounds.

**Comment by collerek** [comment]
Can you provide minimal reproducible examples? Cause in example Mixin has no database representation so how you end up with Mixin in filter?

**Reply by johnthagen** [reply]
Hmm, maybe I was using it incorrectly, but I had a common `Mixin` class that was shared by 3 concrete `Model`s. I wanted to create a function that would take any `Model` that also had that `Mixin`, so I was trying to use `QuerySet[Mixin]` as the type hint.

Maybe this wasn't ever really the correct thing to do. Basically, maybe there could be some discussions that if you want to work with common base layers of Models, you should use inheritance instead of Mixins?

I will try to create an example Monday when I have time to work on this again.

**Reply by johnthagen** [reply]
Here is a minimal example:

```py
import datetime

import databases
import ormar
from ormar import QuerySet
import sqlalchemy

DATABASE_URL = "sqlite:///db.sqlite"
database = databases.Database(DATABASE_URL)
metadata = sqlalchemy.MetaData()


class DateFieldsMixins:
    created_date: datetime.datetime = ormar.DateTime(default=datetime.datetime.now)
    updated_date: datetime.datetime = ormar.DateTime(default=datetime.datetime.now)


class NamedCategory(ormar.Model, DateFieldsMixins):
    class Meta(ormar.ModelMeta):
        tablename = "categories"
        metadata = metadata
        database = database

    id: int = ormar.Integer(primary_key=True)
    name: str = ormar.String(max_length=50, unique=True, index=True)


class CodedCategory(ormar.Model, DateFieldsMixins):
    class Meta(ormar.ModelMeta):
        tablename = "categories"
        metadata = metadata
        database = database

    id: int = ormar.Integer(primary_key=True)
    code: int = ormar.Integer()


def filter_by_created(query: QuerySet[DateFieldsMixins]) -> QuerySet[DateFieldsMixins]:
    return query.filter(DateFieldsMixins.created_date > datetime.datetime.now())


query = filter_by_created(NamedCategory.objects)
...
query = filter_by_created(CodedCategory.objects)
...
```

```bash
$ mypy --ignore-missing-imports main.py
main.py:38: error: Type argument "main4.DateFieldsMixins" of "QuerySet" must be a subtype of "ormar.models.model.Model"
main.py:42: error: Argument 1 to "filter_by_created" has incompatible type "QuerySet[NamedCategory]"; expected "QuerySet[DateFieldsMixins]"
main.py:44: error: Argument 1 to "filter_by_created" has incompatible type "QuerySet[CodedCategory]"; expected "QuerySet[DateFieldsMixins]"
Found 3 errors in 1 file (checked 1 source file)
```

It seems like if you want to work with the common parameters in a generic way, that the concrete table inheritance is the way to go. I just thought if you also agreed, that it might be worth mentioning this in the Inheritance docs.

**Comment by johnthagen** [comment]
Another case where the Pydantic representation clashes with the ORM usage is:

```py
class MyModel(Model):
    length: Optional[str] = Text(nullable=True)


query = MyModel.objects.filter(model_class.length.isnull(False))
# error: Item "str" of "Optional[str]" has no attribute "isnull"
# error: Item "None" of "Optional[str]" has no attribute "isnull
```

I'm not actually sure if there is anything that can be done while preserving ergonomics, but there could be some discussion of this in the user guide so people aren't caught off guard, and perhaps point to alternative syntax if people do not want to add `# type: ignore` lines.
